### PR TITLE
Support None returns from torch.cond

### LINKED ISSUE #181891: torch.cond()'s true_fn and false_fn are not allowed to return None
### 🐛 Describe the bug

torch.cond() currently fails if its true_fn and false_fn return None.

I realized this when I wrote the following code intended to implement loss scaling for mixed precision training inside a cuda graph:

```
import torch
from torch.utils.data import DataLoader, Dataset
from torch._higher_order_ops.cudagraph_conditional_nodes import (
    ControlFlowOpWarmupDispatchMode,
    CUDAGraphCaptureControlFlowOpDispatchMode,
)

class RandomDataset(Dataset):
    def __init__(self, n_samples=256, in_dim=16, out_dim=4):
        self.n_samples = n_samples
        self.in_dim = in_dim
        self.out_dim = out_dim

    def __len__(self):
        return self.n_samples

    def __getitem__(self, _):
        return torch.randn(self.in_dim), torch.randn(self.out_dim)

model = torch.nn.Linear(16, 4).cuda()
criterion = torch.nn.MSELoss()
data_loader = DataLoader(RandomDataset(), batch_size=32, shuffle=False,
                         pin_memory=True, num_workers=1)


lr = torch.tensor(0.1, device="cuda")
loss_scale = torch.tensor(2.0**8, device="cuda")

def train(X, y):
    with torch.autograd.grad_mode.set_multithreading_enabled(False):
        global loss_scale
        model.zero_grad(set_to_none=True)
        y_pred = model(X)
        loss = criterion(y_pred, y)
        (loss * loss_scale).backward()
        all_gradients_finite = torch.stack([
            torch.all(param.grad.isfinite()) for param in model.parameters() if param.grad is not None
        ]).all()
        with torch.no_grad():
            def true_fn():
                global lr, loss_scale
                lr.mul_(0.99999)
                inv_scale = loss_scale.reciprocal()
                for param in model.parameters():
                    if param.grad is not None:
                        param.sub_(lr * param.grad * inv_scale)
                loss_scale.mul_(2.0)

            def false_fn():
                loss_scale.mul_(0.5)

            _ = torch.cond(all_gradients_finite, true_fn, false_fn)
        return loss

for i, (X, y) in enumerate(data_loader):
    if i == 0:
        with ControlFlowOpWarmupDispatchMode():
            loss = train(X.cuda(), y.cuda())
        print(loss.item())
        del loss
        continue
    if i == 1:
        X_static = torch.empty_like(X, device="cuda")
        y_static = torch.empty_like(y, device="cuda")
        g = torch.cuda.CUDAGraph()
        with torch.cuda.graph(g, capture_error_mode="thread_local"), CUDAGraphCaptureControlFlowOpDispatchMode():
            loss_static = train(X_static, y_static)
    if i == 2:
        torch.cuda.cudart().cudaProfilerStart()
    X_static.copy_(X, non_blocking=True)
    y_static.copy_(y, non_blocking=True)
    g.replay()
    print(loss_static.item())
```

The current restriction is a reasonable oversight, since input mutations in torch.cond() were not supported until https://github.com/pytorch/pytorch/pull/172836. Without input mutations, it doesn't make any sense to return nothing. However, things have changed now. You can see clearly that, for gradient nan checks like above, there is no obvious value to return.

I will upload a fix for this shortly. FYI @ydwu4 @kshitij12345 @kiya00 It is likely that torch.while_loop() has a similar issue, though I haven't tested that yet.

### Versions

top of tree

cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @chauhang @ydwu4 @bdhirsh @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #184639

Allow torch.cond branches that only mutate inputs to return None by accepting None constants through Dynamo, CUDA graph conditional capture, and Inductor while guarding unsupported differing constants.

Fixes #181891
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -9997,6 +10002,19 @@ def _require_exact_strides(
...
         assert len(true_outputs) == len(false_outputs), (true_outputs, false_outputs)
         for i, (t_o, f_o) in enumerate(zip(true_outputs, false_outputs)):
+            if isinstance(
+                t_o, (ShapeAsConstantBuffer, NoneAsConstantBuffer)
+            ) or isinstance(f_o, (ShapeAsConstantBuffer, NoneAsConstantBuffer)):
+                assert type(t_o) is type(f_o), (i, t_o, f_o)
+                if isinstance(t_o, ShapeAsConstantBuffer):
+                    assert isinstance(f_o, ShapeAsConstantBuffer)
+                    if t_o.expr != f_o.expr:
+                        raise AssertionError(
```
Comment (aorenste) [inline torch/_inductor/ir.py]: [blocker] `AssertionError` is inconsistent with the cudagraph counterpart added in this same PR (`cudagraph_conditional_nodes.py` consistently uses `ValueError` for the analogous "branches must agree" errors). User-facing capacity limits in Inductor are conventionally reported as `ValueError` or `NotImplementedError`; the message itself reads "Inductor does not support ..." which is exactly the `NotImplementedError` shape. Recommend `raise NotImplementedError(...)` (or `ValueError` to match the sibling) and update the test below to match the specific type.

File: test/inductor/test_control_flow.py [hunk 2]
```diff
@@ -348,6 +348,57 @@ def test_cond_simple_control_flow(self, device, dynamic):
...
+    def test_cond_no_outputs_input_mutation_cpp_wrapper(self):
+        self._check_cond_no_outputs_input_mutation()
+
+    def test_cond_differing_constant_outputs_rejected(self):
+        class Model(torch.nn.Module):
+            def forward(self, p, x):
+                return torch.cond(p, lambda y: 1, lambda y: 2, (x,))
+
+        with self.assertRaisesRegex(
+            Exception,
```
Comment (aorenste) [inline test/inductor/test_control_flow.py]: [cleanup] `assertRaisesRegex(Exception, ...)` is broader than necessary — it'll match any failure that happens to contain the substring "differing constant outputs" (e.g., an unrelated future error that mentions "constant outputs"). Tighten to `AssertionError` (or whichever specific type the paired comment on `ir.py:10012` lands on).

File: torch/_higher_order_ops/cudagraph_conditional_nodes.py [hunk 3]
```diff
@@ -123,8 +123,32 @@ def if_else_node(pred: torch.Tensor, true_fn, false_fn, operands):
...
+                    elif isinstance(if_out, torch.Tensor) or isinstance(
+                        else_out, torch.Tensor
+                    ):
+                        raise ValueError(
+                            "For cudagraph captured torch.cond(), corresponding "
+                            "branch outputs must both be tensors or both be "
+                            "non-tensor constants. "
+                            f"Got {if_out} and {else_out}."
+                        )
+                    elif if_out != else_out:
```
Comment (aorenste) [inline torch/_higher_order_ops/cudagraph_conditional_nodes.py:207]: [cleanup, low-confidence] `if_out != else_out` will falsely raise for NaN-valued Python floats (`nan != nan` is `True`), reporting them as differing constants. The Dynamo gate restricts cond branch constants to `(int, type(None))`, but this dispatch mode is reachable from pure eager code without going through Dynamo (e.g., the new `test_cond_traced_no_outputs_cudagraphs` calls `torch.cond` directly under `CUDAGraphCaptureControlFlowOpDispatchMode`), so a user *could* hit this. Narrow edge case (NaN scalars as cond branch returns are unusual), but worth either (a) restricting allowed constant types to `int`/`None` here too, mirroring the Dynamo gate, or (b) using `not (if_out == else_out)` with a NaN guard.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -9949,13 +9953,14 @@ def create(
 
         def _require_exact_strides(
             graph_outputs: Sequence[IRNode],
-            fake_tensors: Sequence[torch.Tensor],
+            fake_tensors: Sequence[Any],
```
Comment (aorenste) [inline torch/_inductor/ir.py]:
```suggestion
            fake_tensors: Sequence[torch.Tensor | int | None],
```

**Review by aorenste — CHANGES_REQUESTED** [review]

**Comment by zou3519** [comment]
what does JAX's cond allow? is this an exception because we support mutable args?

File: torch/_inductor/ir.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/ir.py]: My agent says I changed the Inductor branch-constant mismatch path in torch/_inductor/ir.py to raise NotImplementedError instead of AssertionError, keeping it as a user-facing unsupported-case error.

File: test/inductor/test_control_flow.py (hunk 2 above)
Comment (jansel) [reply test/inductor/test_control_flow.py]: My agent says I tightened test_cond_differing_constant_outputs_rejected to expect InductorError with the inner NotImplementedError message, instead of catching a broad Exception.

File: torch/_higher_order_ops/cudagraph_conditional_nodes.py (hunk 3 above)
Comment (jansel) [reply torch/_higher_order_ops/cudagraph_conditional_nodes.py:207]: My agent says I mirrored the Dynamo contract in cudagraph_conditional_nodes.py by rejecting non-tensor constants unless they are int or None before value comparison, and added a CUDA graph regression test for NaN constants.

File: torch/_inductor/ir.py (hunk 4 above)
Comment (jansel) [reply torch/_inductor/ir.py]: My agent says I applied the suggested _require_exact_strides annotation, so fake_tensors is now Sequence[torch.Tensor | int | None].

**Comment by jansel** [comment]
My agent says JAX lax.cond does allow both branches to return None; it treats that as an empty pytree result. This PyTorch change is mainly useful because torch.cond now supports input mutation, so a branch can have observable work without returning a tensor. The PR keeps the constant-output contract narrow to tensors plus int/None constants rather than opening arbitrary Python constants.

**Comment by jansel** [comment]
My agent says I checked the current red CI and the representative failed jobs are failing during checkout/submodule fetch with GitHub authentication prompts before any build or test runs. GitHub will not rerun the failed jobs while the workflows are still running, so I am scheduling a follow-up once the run finishes and Dr. CI classifies it.

**Comment by jansel** [comment]
My agent says I reran the failed pull workflow jobs after confirming the two Dr. CI new failures were setup-time GitHub authentication/checkout failures before any build or test code ran. The rerun is now queued, so I am scheduling another follow-up for the new CI result.

**Comment by jansel** [comment]
My agent says I reran the stale inductor and torchtitan workflow checks. The previous red entries were setup/runner failures or stale queued checks, and the PR now has no active failed checks while the reruns are in progress.

**Review by zou3519 — COMMENTED** [review]
claude asked for the following:
```
Recommendation

Request Changes — implementation looks correct (Python-wrapper path verified to already be output-type-agnostic;
constant slots don't shift MultiOutput indices), but add (1) a positive Inductor test for an equal constant output
including cpp_wrapper, and ideally (2) a mixed tensor+None Inductor test and (3) a CUDA-graph pytree-mismatch test.
Confirm the bool-as-int behavior is intended.
```

**Comment by jansel** [comment]
My agent says I addressed the latest review request by adding Inductor coverage for equal constant outputs, including cpp_wrapper, mixed tensor plus None outputs, and CUDA graph pytree mismatch handling. I also tightened the int-or-None constant contract to reject bool explicitly in Dynamo and direct CUDA graph capture, with regression tests for both paths.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and ported the Inductor fix from the removed Conditional IR to the shared Switch IR introduced by #188976. The updated code validates equal int/None constants across every branch, preserves mixed constant/tensor output positions in Python and C++ wrappers, and adds three-branch Switch coverage; the focused Cond, CUDA conditional-node, Switch, and C++ wrapper tests pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the July 27 review and the related Ruff CI failure. Cond, switch, and direct CUDA capture now share the exact int-or-None contract with an explicit bool error; graph-break history is preserved; the lint-failing asserts use explicit AssertionError raises; mixed None/type leaves fail clearly at fake merge; the new Inductor tests cover CPU/CUDA, static/dynamic, and Python/C++ wrappers; and CUDA constant capture now exercises replay. The focused 40-case Inductor matrix, 7 CUDA conditional-node tests, and full lintrunner all pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the July 31 review: the public cond/switch docs now describe int/None outputs and the Inductor limitation; user-facing type mismatches no longer surface as internal assertions; the switch and C++ sentinel contracts are documented; bool contract tests moved out of the Inductor matrix; cond now has mixed tensor/None C++ wrapper coverage; and CUDA constant-output replay observes device state across both predicates. The 21 focused Dynamo, Inductor, and CUDA tests plus full lintrunner pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
